examples/hmm.py

```diff
@@ -0,0 +1,179 @@
+    # inject log_prob to potential function
+    return sample('forward_log_prob', dist.Multinomial(logits=-log_prob), obs=0)
+
+
+def run_inference(transition_prior, emission_prior, supervised_categories, supervised_words,
+                  unsupervised_words, rng, args):
+    init_params, potential_fn, transform_fn = initialize_model(
+        rng,
+        semi_supervised_hmm,
+        transition_prior, emission_prior, supervised_categories,
+        supervised_words, unsupervised_words,
+    )
+    init_kernel, sample_kernel = hmc(potential_fn, algo='NUTS')
+    hmc_state = init_kernel(init_params, args.num_warmup)
+    hmc_states = fori_collect(args.num_samples, sample_kernel, hmc_state,
+                              transform=lambda state: transform_fn(state.z))
+    return hmc_states
+
+
+def print_results(posterior, transition_prob, emission_prob):
+    header = semi_supervised_hmm.__name__ + ' - TRAIN'
+    columns = ['', 'ActualProb', 'Pred(p25)', 'Pred(p50)', 'Pred(p75)']
+    header_format = '{:>20} {:>10} {:>10} {:>10} {:>10}'
+    row_format = '{:>20} {:>10.2f} {:>10.2f} {:>10.2f} {:>10.2f}'
+    print('\n', '=' * 20 + header + '=' * 20, '\n')
+    print(header_format.format(*columns))
+
+    quantiles = onp.quantile(posterior['transition_prob'], [0.25, 0.5, 0.75], axis=0)
+    for i in range(transition_prob.shape[0]):
+        for j in range(transition_prob.shape[1]):
+            idx = 'transition[{},{}]'.format(i, j)
+            print(row_format.format(idx, transition_prob[i, j], *quantiles[:, i, j]), '\n')
+
+    quantiles = onp.quantile(posterior['emission_prob'], [0.25, 0.5, 0.75], axis=0)
+    for i in range(emission_prob.shape[0]):
+        for j in range(emission_prob.shape[1]):
+            idx = 'emission[{},{}]'.format(i, j)
+            print(row_format.format(idx, emission_prob[i, j], *quantiles[:, i, j]), '\n')
+
+
+def main(args):
+    jax_config.update('jax_platform_name', args.device)
+    print('Simulating data...')
+    (transition_prior, emission_prior, transition_prob, emission_prob,
+     supervised_categories, supervised_words, unsupervised_words) = simulate_data(
+        random.PRNGKey(1),
+        num_categories=args.num_categories,
+        num_words=args.num_words,
+        num_supervised_data=args.num_supervised,
+        num_unsupervised_data=args.num_unsupervised,
+    )
+    print('Starting inference...')
+    zs = run_inference(transition_prior, emission_prior,
+                       supervised_categories, supervised_words, unsupervised_words,
+                       random.PRNGKey(2), args)
+    print_results(zs, transition_prob, emission_prob)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Semi-supervised Hidden Markov Model')
+    parser.add_argument('--num-categories', default=3, type=int)
+    parser.add_argument('--num-words', default=10, type=int)
+    parser.add_argument('--num-supervised', default=100, type=int)
```

Comment:
Just for my understanding, how low can we go here before inference fails completely?

Reply:
![hmm20](https://user-images.githubusercontent.com/4736342/57737619-4d828d80-767a-11e9-8b2c-20aefaa40ad3.png)
![hmm-10](https://user-images.githubusercontent.com/4736342/57737620-4d828d80-767a-11e9-86e5-e0fdffa133ac.png)
FYI, the first one is `num_supervised=10`, the second one is `num_supervised=2`.

Reply:
So I guess it fails completely when n-sup between 2-10.

Reply:
Interesting, thanks for the plots! The distributions don't look that much more different and the trends are kind of preserved which is good, but I wonder if we would end up with different results for other chains.